### PR TITLE
AzureMonitor: Use auth middleware for QueryData requests

### LINKED ISSUE #35027: AzureMonitor: Use auth middleware for QueryData requests

### PR DESCRIPTION
**Pull request by andresmgot**
**What this PR does / why we need it**:

The goal of this PR is to avoid to use the plugin proxy to generate data requests and use an authentication middleware instead. It's also one of the purposes of this PR to split the code related to the retrieval of the auth token to its own package so it can be reused for different Azure plugins.

**Which issue(s) this PR fixes**:

Fixes #35027

**Special notes for your reviewer**:

Note that `pkg/tsdb/azuremonitor/tokenprovider/` is just a copy of the code (with some minimal adjustments) from ` pkg/api/pluginproxy/`.

For the moment, the `routes` are duplicated between this file and `plugin.json` but the goal is to remove the routes from `plugin.json` once the frontend stop using them (which won't happen in this PR).

### CODE REVIEW

File: pkg/tsdb/azuremonitor/credentials.go [hunk 1]
```diff
@@ -1,15 +1,20 @@
...
+	"github.com/grafana/grafana/pkg/plugins"
 	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/tsdb/azuremonitor/tokenprovider"
 )
 
-const (
-	AzureAuthManagedIdentity = "msi"
-	AzureAuthClientSecret    = "clientsecret"
-)
+// TODO: Remove?
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/credentials.go:13]: Yes!

File: pkg/tsdb/azuremonitor/tokenprovider/token_provider_azure.go [hunk 2]
```diff
@@ -18,27 +17,19 @@ var (
...
-	ctx               context.Context
-	cfg               *setting.Cfg
-	route             *plugins.AppPluginRoute
-	authParams        *plugins.JwtTokenAuth
+	cfg        *setting.Cfg
+	authParams *plugins.JwtTokenAuth
 }
 
-func newAzureAccessTokenProvider(ctx context.Context, cfg *setting.Cfg, ds *models.DataSource, pluginRoute *plugins.AppPluginRoute,
+func NewAzureAccessTokenProvider(cfg *setting.Cfg,
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/tokenprovider/token_provider_azure.go:24]: You should still be able to pass context to this func. You can use the request context within the auth middleware.

File: pkg/tsdb/azuremonitor/tokenprovider/token_provider_azure.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package pluginproxy
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/tokenprovider/token_provider_azure.go:1]: Can we call this just token_provider.go? The azure prefix is given by the context I think.

File: pkg/tsdb/azuremonitor/tokenprovider/token_provider.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package tokenprovider
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/tokenprovider/token_provider.go:1]: And maybe this can be called authentication_middleware?

**Review by sunker — COMMENTED** [review]
Really happy to see this refactor in coming through. This is such an improvement! Can't wait to see what it's going to look like once call resource handler is also implemented. 

I like the idea of having routes defined the way you've done it in routes.go. 

Just added a few minor comments. Haven't tested this yet but I will as soon as the PR is moved from draft state.

File: pkg/tsdb/azuremonitor/tokenprovider/token_provider_azure.go (hunk 2 above)
Comment (andresmgot) [reply pkg/tsdb/azuremonitor/tokenprovider/token_provider_azure.go:24]: moved the tokenprovider initialization to the request context to be able to use its ctx

**Comment by andresmgot** [comment]
This should be ready to be reviewed! (I still need to do some IRL testing for az resource graph and insights analytics).

**Comment by andresmgot** [comment]
> (I still need to do some testing IRL for az resource graph and insights analytics).

Done! (and fixed)

**Review by sunker — APPROVED** [review]
Great work on this!

File: pkg/tsdb/azuremonitor/credentials.go [hunk 5]
```diff
@@ -19,59 +18,40 @@ const (
...
+	azureResourceGraph = "Azure Resource Graph"
+)
 
-		// For newly created datasource with no configuration, managed identity is the default authentication type
-		// if they are enabled in Grafana config
-		if cfg.Azure.ManagedIdentityEnabled {
-			return AzureAuthManagedIdentity
-		} else {
-			return AzureAuthClientSecret
+func httpCliProvider(ctx context.Context, route azRoute, model datasourceInfo, cfg *setting.Cfg) *httpclient.Provider {
```
[the remark this answers is not in the document]
Comment (andresmgot) [reply pkg/tsdb/azuremonitor/credentials.go]: that's 3 letters more, we are going to get charged by the extra disk space used :P

File: pkg/tsdb/azuremonitor/credentials.go [hunk 6]
```diff
@@ -19,59 +18,40 @@ const (
...
+	if len(route.Scopes) > 0 {
```
[the remark this answers is not in the document]
Comment (andresmgot) [reply pkg/tsdb/azuremonitor/credentials.go:31]: That's what I thought but App Insights and Insights Analytics don't have a scope. They use an API key in some header: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/grafana-azure-monitor-datasource/plugin.json#L105

**Pull request merged**